File: next/security/backend-ms1/src/main/java/ma/zs/stocky/ws/facade/admin/security/RoleRestAdmin.java
package  ma.zs.stocky.ws.facade.admin.security;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;

import ma.zs.stocky.bean.core.security.Role;
import ma.zs.stocky.dao.criteria.core.security.RoleCriteria;
import ma.zs.stocky.service.facade.admin.security.RoleAdminService;
import ma.zs.stocky.ws.converter.security.RoleConverter;
import ma.zs.stocky.ws.dto.security.RoleDto;
import ma.zs.stocky.zynerator.controller.AbstractController;
import ma.zs.stocky.zynerator.dto.AuditEntityDto;
import ma.zs.stocky.zynerator.util.PaginatedList;


import org.springframework.core.io.InputStreamResource;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;


import java.util.List;
import org.springframework.beans.factory.annotation.Autowired;
import ma.zs.stocky.zynerator.process.Result;


import org.springframework.web.multipart.MultipartFile;
import ma.zs.stocky.zynerator.dto.FileTempDto;

@RestController
@RequestMapping("/api/admin/role/")
public class RoleRestAdmin  extends AbstractController<Role, RoleDto, RoleCriteria, RoleAdminService, RoleConverter> {



    @Operation(summary = "upload one role")
    @RequestMapping(value = "upload", method = RequestMethod.POST, consumes = "multipart/form-data")
    public ResponseEntity<FileTempDto> uploadFileAndGetChecksum(@RequestBody MultipartFile file) throws Exception {
        return super.uploadFileAndGetChecksum(file);
    }
    @Operation(summary = "upload multiple roles")
    @RequestMapping(value = "upload-multiple", method = RequestMethod.POST, consumes = "multipart/form-data")
    public ResponseEntity<List<FileTempDto>> uploadMultipleFileAndGetChecksum(@RequestBody MultipartFile[] files) throws Exception {
        return super.uploadMultipleFileAndGetChecksum(files);
    }

    @Operation(summary = "Finds a list of all roles")
    @GetMapping("")
    public ResponseEntity<List<RoleDto>> findAll() throws Exception {
        return super.findAll();
    }

    @Operation(summary = "Finds an optimized list of all roles")
    @GetMapping("optimized")
    public ResponseEntity<List<RoleDto>> findAllOptimized() throws Exception {
        return super.findAllOptimized();
    }

    @Operation(summary = "Finds a role by authority")
    @GetMapping("authority/{authority}")
    public ResponseEntity<RoleDto> findByAuthority(@PathVariable String authority) {
        return super.findByReferenceEntity(new Role(authority));
    }

    @Operation(summary = "Saves the specified  role")
    @PostMapping("")
    public ResponseEntity<RoleDto> save(@RequestBody RoleDto dto) throws Exception {
        return super.save(dto);
    }

    @Operation(summary = "Updates the specified  role")
    @PutMapping("")
    public ResponseEntity<RoleDto> update(@RequestBody RoleDto dto) throws Exception {
        return super.update(dto);
    }

    @Operation(summary = "Delete list of role")
    @PostMapping("multiple")
    public ResponseEntity<List<RoleDto>> delete(@RequestBody List<RoleDto> listToDelete) throws Exception {
        return super.delete(listToDelete);
    }
    @Operation(summary = "Delete the specified role")
    @DeleteMapping("")
    public ResponseEntity<RoleDto> delete(@RequestBody RoleDto dto) throws Exception {
            return super.delete(dto);
    }

    @Operation(summary = "Delete the specified role")
    @DeleteMapping("id/{id}")
    public ResponseEntity<Long> deleteById(@PathVariable Long id) throws Exception {
        return super.deleteById(id);
    }
    @Operation(summary = "Delete multiple roles by ids")
    @DeleteMapping("multiple/id")
    public ResponseEntity<List<Long>> deleteByIdIn(@RequestBody List<Long> ids) throws Exception {
            return super.deleteByIdIn(ids);
     }



    @Operation(summary = "Finds a role and associated list by id")
    @GetMapping("id/{id}")
    public ResponseEntity<RoleDto> findById(@PathVariable Long id) {
        return super.findById(id);
    }

    @Operation(summary = "Finds roles by criteria")
    @PostMapping("find-by-criteria")
    public ResponseEntity<List<RoleDto>> findByCriteria(@RequestBody RoleCriteria criteria) throws Exception {
        return super.findByCriteria(criteria);
    }

    @Operation(summary = "Finds paginated roles by criteria")
    @PostMapping("find-paginated-by-criteria")
    public ResponseEntity<PaginatedList> findPaginatedByCriteria(@RequestBody RoleCriteria criteria) throws Exception {
        return super.findPaginatedByCriteria(criteria);
    }

    @Operation(summary = "Exports roles by criteria")
    @PostMapping("export")
    public ResponseEntity<InputStreamResource> export(@RequestBody RoleCriteria criteria) throws Exception {
        return super.export(criteria);
    }

    @Operation(summary = "Gets role data size by criteria")
    @PostMapping("data-size-by-criteria")
    public ResponseEntity<Integer> getDataSize(@RequestBody RoleCriteria criteria) throws Exception {
        return super.getDataSize(criteria);
    }



    public RoleRestAdmin (RoleAdminService service, RoleConverter converter) {
        super(service, converter);
    }




}

File: nest-angular/security/backend-ms1/src/app/module/ws/admin/security/RoleRestAdmin.ts
import {Body, Controller,Delete , Get, Param, Post, Put} from '@nestjs/common';
import {ApiOperation, ApiTags} from '@nestjs/swagger';
import {PaginatedList} from "src/app/zynerator/util/PaginatedList";

import {RoleConverter} from "src/app/shared/converter/security/RoleConverter";
import {RoleDto} from "src/app/shared/dto/security/RoleDto";
import {RoleCriteria} from "src/app/shared/dao/criteria/core/security/RoleCriteria";
import {RoleAdminServiceImpl} from "src/app/module/service/admin/security/RoleAdminServiceImpl";


@ApiTags('Manages role services')
@Controller('api/admin/role')
export class RoleRestAdmin {

    constructor(private readonly service: RoleAdminServiceImpl,
                private readonly converter: RoleConverter) {
    }

    @ApiOperation({summary: 'Finds a list of all roles'})
    @Get()
    async findAll(): Promise<RoleDto[]> {
        const items = await this.service.findAll();
        return this.converter.toDtos(items);
    }

    @ApiOperation({summary: 'Finds a role by id'})
    @Get('id/:id')
    async findById(@Param('id') id: number): Promise<RoleDto> {
        const item = await this.service.findById(id);
        return this.converter.toDto(item);
    }

    @ApiOperation({summary: 'Deletes a role by id'})
    @Delete('id/:id')
    async deleteById(@Param('id') id: number): Promise<number> {
        return this.service.deleteById(id);
    }

    @ApiOperation({summary: 'Saves the specified role'})
    @Post()
    async save(@Body() dto: RoleDto): Promise<RoleDto> {
        const item = this.converter.toItem(dto);
        const savedItem = await this.service.save(item);
        return this.converter.toDto(savedItem);
    }


    @ApiOperation({summary: 'Updates the specified role'})
    @Put()
    async update(@Body() dto: RoleDto): Promise<RoleDto> {
        const item = this.converter.toItem(dto);
        const result = await this.service.updateWithAssociatedLists(item);
        return this.converter.toDto(result);
   }

    @ApiOperation({summary: 'Finds an optimized list of all roles'})
    @Get('optimized')
    async findAllOptimized(): Promise<RoleDto[]> {
        const result = await this.service.findAllOptimized();
        return result;
    }

    @ApiOperation({summary: 'Finds an optimized list of all roles'})
    @Post('find-by-criteria')
    async findByCriteria(@Body() criteria: RoleCriteria): Promise<RoleDto[]> {
        const items = await this.service.findByCriteria(criteria);
        return this.converter.toDtos(items);
    }

    @ApiOperation({summary: 'Finds an optimized list of all roles'})
    @Post('find-paginated-by-criteria')
    async findPaginatedByCriteria(@Body() criteria: RoleCriteria): Promise<PaginatedList<RoleDto>> {
        const paginated = await this.service.findPaginatedByCriteria(criteria);
        const dtos = this.converter.toDtos(paginated.list);
        return new PaginatedList<RoleDto>(dtos, paginated.dataSize);
    }



    @Post('multiple')
    async deleteMultiple(@Body() dtos: RoleDto[]): Promise<RoleDto[]> {
        const items = dtos.map(dto => this.converter.toItem(dto));
        const deletedItems = await this.service.deleteMultiple(items);
        return deletedItems.map(element => this.converter.toDto(element));
    }

}
